sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala

```diff
@@ -228,34 +227,79 @@ class GpuSorter(
+   *
+   * After this function is called, the argument `spillableBatches` should not be used.
+   *
+   * @param spillableBatches the spillable batches to sort
    * @param sortTime metric for the time spent doing the merge sort
    * @return the sorted data.
    */
-  final def mergeSort(batches: Array[ColumnarBatch], sortTime: GpuMetric): ColumnarBatch = {
+  final def mergeSortAndClose(
+      spillableBatches: mutable.ArrayStack[SpillableColumnarBatch],
+      sortTime: GpuMetric): ColumnarBatch = {
     withResource(new NvtxWithMetrics("merge sort", NvtxColor.DARK_GREEN, sortTime)) { _ =>
-      if (batches.length == 1) {
-        GpuColumnVector.incRefCounts(batches.head)
+      if (spillableBatches.size == 1) {
+        // Single batch no need for a merge sort
+        withResource(spillableBatches.pop()) { sb =>
+          sb.getColumnarBatch()
+        }
       } else {
-        val merged = withResource(ArrayBuffer[Table]()) { tabs =>
-          batches.foreach { cb =>
-            tabs += GpuColumnVector.from(cb)
-          }
-          // In the current version of cudf merge does not work for lists and maps.
-          // This should be fixed by https://github.com/rapidsai/cudf/issues/8050
-          // Nested types in sort key columns is not supported either.
-          if (hasNestedInKeyColumns || hasUnsupportedNestedInRideColumns) {
-            // so as a work around we concatenate all of the data together and then sort it.
-            // It is slower, but it works
-            withResource(Table.concatenate(tabs: _*)) { concatenated =>
-              concatenated.orderBy(cudfOrdering: _*)
+        closeOnExcept(spillableBatches) { _ =>
+          val merged = {
+            val tablesToMerge = new mutable.ArrayStack[Table]()
+            while (spillableBatches.nonEmpty || tablesToMerge.size > 1) {
+              // pop a spillable batch if there is one to pop
+              var spillableCandidate: Option[SpillableColumnarBatch] = None
+              if (spillableBatches.nonEmpty) {
+                spillableCandidate = Some(spillableBatches.pop())
+              }
+
+              // optionally add a table to `tablesToMerge`
+              closeOnExcept(tablesToMerge) { _ =>
+                withResource(spillableCandidate) {
+                  _.foreach { sb =>
+                    // materialize a potentially spilled batch
+                    withResource(sb.getColumnarBatch()) { batch =>
+                      tablesToMerge.push(GpuColumnVector.from(batch))
+                    }
+                  }
+                }
+              }
+
+              if (tablesToMerge.size > 1) {
+                // In the current version of cudf merge does not work for lists and maps.
+                // This should be fixed by https://github.com/rapidsai/cudf/issues/8050
+                // Nested types in sort key columns is not supported either.
+                if (hasNestedInKeyColumns || hasUnsupportedNestedInRideColumns) {
```

Comment:
This corner case is common enough that I think we should optimize this a bit.  A full sort is a lot more expensive than a merge sort, and now we are doing `N-1` full sorts where as before we were only doing 1 full sort.  Could we move this up so the code is more like

```
if (spillableBatches.size == 1) {
} else if (hasNestedInKeyColumns || hasUnsupportedNestedInRideColumns) {
  // Unspill all of the input batches
  // concat the input batches
  // close the input batches
  // sort the concated batch
  // close the concated batch
} else {
  ...
}
```

Reply:
@revans2 that should be fixed now